fecgov/openFEC
# candidate totals by party by office

## Summary (required)

- Resolves #3549 Make new /candidates/totals/by_party/by_office endpoint

Data Source:
ofec_candidate_totals_mv has the financial information aggregated by candidate/election_cycles available already. It does not have the office, party, and candidate_active information though.
However, ofec_candidate_history_with_future_election_vw is the driver in the main query,
which already has all the additional candidate information needed:
office
party
candidate_active
Add these 3 fields in the list of select.

new endpoint named /candidates/totals/by_office/by_party/ had been created.

## How to test the changes locally
A tmp version of the updated mv (ofec_candidate_totals_mv_tmp_jj) had been created in DEV database. 

Connected to the DEV database and execute the following query, it should show the total receipts and total disbursement for all active candidates of election_year 2020 for office= 'P': 
SELECT ofec_candidate_totals_mv_tmp_jj.office AS office, 
CASE WHEN (ofec_candidate_totals_mv_tmp_jj.party = 'DFL') THEN 'DEM' WHEN (ofec_candidate_totals_mv_tmp_jj.party = 'DEM') THEN 'DEM' WHEN (ofec_candidate_totals_mv_tmp_jj.party = 'REP') THEN 'REP' ELSE 'Other' END AS party,
 ofec_candidate_totals_mv_tmp_jj.election_year AS election_year, 
 sum(ofec_candidate_totals_mv_tmp_jj.receipts) AS total_receipts, 
 sum(ofec_candidate_totals_mv_tmp_jj.disbursements) AS total_disbursements 
FROM ofec_candidate_totals_mv_tmp_jj 
WHERE ofec_candidate_totals_mv_tmp_jj.is_election = true AND ofec_candidate_totals_mv_tmp_jj.office = 'P' AND ofec_candidate_totals_mv_tmp_jj.election_year IN (2020) 
AND ofec_candidate_totals_mv_tmp_jj.candidate_inactive = False
GROUP BY ofec_candidate_totals_mv_tmp_jj.office, 
CASE WHEN (ofec_candidate_totals_mv_tmp_jj.party = 'DFL') THEN 'DEM' WHEN (ofec_candidate_totals_mv_tmp_jj.party = 'DEM') THEN 'DEM' WHEN (ofec_candidate_totals_mv_tmp_jj.party = 'REP') THEN 'REP' ELSE 'Other' END, ofec_candidate_totals_mv_tmp_jj.election_year; 


Download this branch to local machine. 
In the model file candidates.py, change the following line to point to the tmp mv
class CandidateTotal(db.Model):
    __tablename__ = 'ofec_candidate_totals_mv_tmp_jj'

Start the local server. 
Under candidate, looking for the new endpoint
/candidates/totals/by_office/by_party/ 

Tested with election_year/office/party.  
2020/P/true (the result should match the query above)
2020/S/true (you can modify the query above to verify the result)
or other combinations you prefer.

Also check is_active_candidate
if no value specified, all candidates data is returned (in the query, AND ofec_candidate_totals_mv_tmp_jj.candidate_inactive = True/False is omitted). 
When True is specified, only active candidates data are returned (in the query, candidate_inactive = False). 
When False is specified, only inactive candidates data is returned (in the query, candidate_inactive = True).

## Impacted areas of the application
List general components of the application that this PR will affect:

New endpoint under "candidate"
/candidates/totals/by_office/by_party/ 


## Related PRs
List related PRs against other branches:

branch | PR
------ | ------
fix/other_pr | [link]()
feature/other_pr | [link]()
